File: goldenpath/canary/judge.py
"""Kayenta-style automated canary judgment.

Mirrors the shape of Netflix's NetflixACAJudge: each configured metric is
classified independently against the baseline, then the classifications are
combined into a single weighted score which is compared against pass and
marginal thresholds.

Two guardrails matter more than the score itself:

* A metric may only fail if the difference is *both* statistically significant
  and large enough to care about. Significance alone is not enough -- with
  enough samples, a 0.3ms latency regression is significant and irrelevant.
* A metric marked `critical` fails the whole canary on its own, no matter how
  healthy the weighted score looks.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field

from .stats import MannWhitneyResult, mann_whitney_u, percentile


class Classification(enum.StrEnum):
    """Per-metric outcome, using Kayenta's vocabulary."""

    PASS = "Pass"
    HIGH = "High"  # Canary significantly higher than baseline, and that is bad.
    LOW = "Low"  # Canary significantly lower than baseline, and that is bad.
    NODATA = "Nodata"
    ERROR = "Error"

    @property
    def is_failure(self) -> bool:
        return self in (Classification.HIGH, Classification.LOW, Classification.ERROR)


class Verdict(enum.StrEnum):
    PASS = "PASS"
    MARGINAL = "MARGINAL"
    FAIL = "FAIL"


class Direction(enum.StrEnum):
    """Which way a metric moving is considered a regression."""

    INCREASE = "increase"  # Only an increase is bad (latency, error rate).
    DECREASE = "decrease"  # Only a decrease is bad (throughput, cache hits).
    EITHER = "either"  # Any significant change is suspicious.


@dataclass(frozen=True)
class MetricSpec:
    """How one metric should be judged.

    Attributes:
        name: Metric identifier, matched against collected sample keys.
        direction: Which direction constitutes a regression.
        weight: Relative contribution to the aggregate score.
        critical: If true, a failure here fails the canary outright.
        tolerance: Minimum |Cliff's delta| for a significant difference to
            count as a real regression. Guards against statistically real but
            operationally meaningless changes.
        significance: Alpha for the Mann-Whitney test.
        min_samples: Below this, the metric is classified Nodata.
    """

    name: str
    direction: Direction = Direction.INCREASE
    weight: float = 1.0
    critical: bool = False
    tolerance: float = 0.2
    significance: float = 0.05
    min_samples: int = 20


@dataclass
class MetricResult:
    name: str
    classification: Classification
    reason: str
    weight: float
    critical: bool
    canary_count: int = 0
    baseline_count: int = 0
    canary_median: float | None = None
    baseline_median: float | None = None
    canary_p95: float | None = None
    baseline_p95: float | None = None
    p_value: float | None = None
    cliffs_delta: float | None = None

    @property
    def passed(self) -> bool:
        return not self.classification.is_failure

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "classification": self.classification.value,
            "reason": self.reason,
            "weight": self.weight,
            "critical": self.critical,
            "canary_count": self.canary_count,
            "baseline_count": self.baseline_count,
            "canary_median": self.canary_median,
            "baseline_median": self.baseline_median,
            "canary_p95": self.canary_p95,
            "baseline_p95": self.baseline_p95,
            "p_value": self.p_value,
            "cliffs_delta": self.cliffs_delta,
        }


@dataclass
class CanaryResult:
    """Aggregate judgment across every metric."""

    score: float
    verdict: Verdict
    results: list[MetricResult] = field(default_factory=list)
    pass_threshold: float = 95.0
    marginal_threshold: float = 75.0
    summary: str = ""

    @property
    def promotable(self) -> bool:
        return self.verdict is Verdict.PASS

    @property
    def failures(self) -> list[MetricResult]:
        return [r for r in self.results if not r.passed]

    def to_dict(self) -> dict:
        return {
            "score": round(self.score, 2),
            "verdict": self.verdict.value,
            "pass_threshold": self.pass_threshold,
            "marginal_threshold": self.marginal_threshold,
            "summary": self.summary,
            "metrics": [r.to_dict() for r in self.results],
        }


def _median(values: list[float]) -> float:
    return percentile(values, 50.0)


def classify_metric(
    spec: MetricSpec,
    canary: list[float],
    baseline: list[float],
) -> MetricResult:
    """Classify a single metric's canary sample against its baseline."""
    base = MetricResult(
        name=spec.name,
        classification=Classification.NODATA,
        reason="",
        weight=spec.weight,
        critical=spec.critical,
        canary_count=len(canary),
        baseline_count=len(baseline),
    )

    if len(canary) < spec.min_samples or len(baseline) < spec.min_samples:
        base.reason = (
            f"insufficient samples (canary={len(canary)}, baseline={len(baseline)}, "
            f"required={spec.min_samples})"
        )
        return base

    base.canary_median = _median(canary)
    base.baseline_median = _median(baseline)
    base.canary_p95 = percentile(canary, 95.0)
    base.baseline_p95 = percentile(baseline, 95.0)

    test: MannWhitneyResult = mann_whitney_u(canary, baseline)
    base.p_value = test.p_value
    base.cliffs_delta = test.cliffs_delta

    significant = test.p_value < spec.significance
    material = abs(test.cliffs_delta) >= spec.tolerance

    if not significant:
        base.classification = Classification.PASS
        base.reason = f"no significant difference (p={test.p_value:.4f})"
        return base

    if not material:
        base.classification = Classification.PASS
        base.reason = (
            f"significant but within tolerance "
            f"(p={test.p_value:.4f}, delta={test.cliffs_delta:+.3f}, "
            f"tolerance={spec.tolerance})"
        )
        return base

    canary_is_higher = test.cliffs_delta > 0
    regression = (
        (spec.direction is Direction.INCREASE and canary_is_higher)
        or (spec.direction is Direction.DECREASE and not canary_is_higher)
        or (spec.direction is Direction.EITHER)
    )

    if not regression:
        base.classification = Classification.PASS
        base.reason = (
            f"moved in the improving direction "
            f"(delta={test.cliffs_delta:+.3f}, p={test.p_value:.4f})"
        )
        return base

    base.classification = Classification.HIGH if canary_is_higher else Classification.LOW
    base.reason = (
        f"canary {'higher' if canary_is_higher else 'lower'} than baseline: "
        f"median {base.canary_median:.2f} vs {base.baseline_median:.2f}, "
        f"p95 {base.canary_p95:.2f} vs {base.baseline_p95:.2f} "
        f"(p={test.p_value:.4f}, delta={test.cliffs_delta:+.3f})"
    )
    return base


def judge(
    specs: list[MetricSpec],
    canary_samples: dict[str, list[float]],
    baseline_samples: dict[str, list[float]],
    pass_threshold: float = 95.0,
    marginal_threshold: float = 75.0,
) -> CanaryResult:
    """Judge a canary across every configured metric.

    Nodata metrics are excluded from the score rather than counted as passes:
    a metric that reported nothing is an unknown, and silently treating an
    unknown as healthy is how bad builds reach production.

    Returns:
        A CanaryResult whose verdict is PASS, MARGINAL, or FAIL.
    """
    if not specs:
        raise ValueError("canary judgment requires at least one metric spec")

    results = [
        classify_metric(
            spec,
            canary_samples.get(spec.name, []),
            baseline_samples.get(spec.name, []),
        )
        for spec in specs
    ]

    scored = [r for r in results if r.classification is not Classification.NODATA]
    if not scored:
        return CanaryResult(
            score=0.0,
            verdict=Verdict.FAIL,
            results=results,
            pass_threshold=pass_threshold,
            marginal_threshold=marginal_threshold,
            summary="No metric produced enough data to judge; refusing to promote.",
        )

    total_weight = sum(r.weight for r in scored)
    passing_weight = sum(r.weight for r in scored if r.passed)
    score = 100.0 * passing_weight / total_weight if total_weight else 0.0

    critical_failures = [r for r in results if not r.passed and r.critical]

    if critical_failures:
        verdict = Verdict.FAIL
        names = ", ".join(r.name for r in critical_failures)
        summary = f"Critical metric regression: {names}. Score {score:.1f} is not consulted."
    elif score >= pass_threshold:
        verdict = Verdict.PASS
        summary = f"Score {score:.1f} >= pass threshold {pass_threshold}."
    elif score >= marginal_threshold:
        verdict = Verdict.MARGINAL
        summary = (
            f"Score {score:.1f} sits between marginal ({marginal_threshold}) "
            f"and pass ({pass_threshold}) thresholds; human judgment required."
        )
    else:
        verdict = Verdict.FAIL
        summary = f"Score {score:.1f} < marginal threshold {marginal_threshold}."

    nodata = [r for r in results if r.classification is Classification.NODATA]
    if nodata:
        summary += f" Excluded {len(nodata)} metric(s) with insufficient data."

    return CanaryResult(
        score=score,
        verdict=verdict,
        results=results,
        pass_threshold=pass_threshold,
        marginal_threshold=marginal_threshold,
        summary=summary,
    )

File: docs/site/judge.js
/*
 * A faithful JavaScript port of goldenpath/canary/judge.py and stats.py.
 *
 * This exists so the canary judge can be explored interactively in a browser
 * without a server. It is a port, not a reimplementation: the same
 * Mann-Whitney U test, the same tie and continuity corrections, the same
 * Cliff's delta, the same tolerance gate, the same weighted score.
 *
 * Ports drift. `tests/test_js_parity.py` runs this file under Node against
 * fixed inputs and asserts the numbers match the Python judge -- exactly for
 * the rank statistics, and to 1e-6 for p-values, which is the floor set by the
 * error-function approximation below. If the two ever diverge, CI says so.
 */
(function (root) {
  'use strict';

  // ---------------------------------------------------------------- stats --

  function fractionalRanks(values) {
    const order = values.map((v, i) => i).sort((a, b) => values[a] - values[b]);
    const ranks = new Array(values.length).fill(0);
    let i = 0;
    while (i < order.length) {
      let j = i;
      while (j + 1 < order.length && values[order[j + 1]] === values[order[i]]) j++;
      const midpoint = (i + j + 2) / 2;
      for (let k = i; k <= j; k++) ranks[order[k]] = midpoint;
      i = j + 1;
    }
    return ranks;
  }

  function tieCorrectionTerm(values) {
    const ordered = values.slice().sort((a, b) => a - b);
    let total = 0;
    let i = 0;
    while (i < ordered.length) {
      let j = i;
      while (j + 1 < ordered.length && ordered[j + 1] === ordered[i]) j++;
      const t = j - i + 1;
      total += t * t * t - t;
      i = j + 1;
    }
    return total;
  }

  // Abramowitz & Stegun 7.1.26; accurate to ~1.5e-7, well inside the
  // tolerance the parity test enforces on the resulting p-values.
  function erf(x) {
    const sign = x < 0 ? -1 : 1;
    const ax = Math.abs(x);
    const t = 1 / (1 + 0.3275911 * ax);
    const y =
      1 -
      ((((1.061405429 * t - 1.453152027) * t + 1.421413741) * t - 0.284496736) * t +
        0.254829592) *
        t *
        Math.exp(-ax * ax);
    return sign * y;
  }

  function normalCdf(z) {
    return 0.5 * (1 + erf(z / Math.SQRT2));
  }

  function percentile(values, q) {
    if (!values.length) throw new Error('percentile of an empty sample is undefined');
    const ordered = values.slice().sort((a, b) => a - b);
    if (ordered.length === 1) return ordered[0];
    const position = (ordered.length - 1) * (q / 100);
    const lower = Math.floor(position);
    const upper = Math.ceil(position);
    if (lower === upper) return ordered[position];
    return ordered[lower] * (1 - (position - lower)) + ordered[upper] * (position - lower);
  }

  function mannWhitneyU(canary, baseline) {
    const n1 = canary.length;
    const n2 = baseline.length;
    if (!n1 || !n2) throw new Error('Mann-Whitney U requires two non-empty samples');

    const combined = canary.concat(baseline);
    const ranks = fractionalRanks(combined);

    let rankSum = 0;
    for (let i = 0; i < n1; i++) rankSum += ranks[i];
    const u = rankSum - (n1 * (n1 + 1)) / 2;
    const cliffsDelta = (2 * u) / (n1 * n2) - 1;

    const total = n1 + n2;
    const meanU = (n1 * n2) / 2;
    const tieTerm = tieCorrectionTerm(combined);
    const variance =
      total < 2 ? 0 : ((n1 * n2) / 12) * (total + 1 - tieTerm / (total * (total - 1)));

    if (variance <= 0) {
      return { uStatistic: u, zScore: 0, pValue: 1, cliffsDelta: cliffsDelta };
    }

    const sigma = Math.sqrt(variance);
    let numerator = Math.abs(u - meanU) - 0.5;
    if (numerator < 0) numerator = 0;
    let z = numerator / sigma;
    if (u < meanU) z = -z;

    let p = 2 * (1 - normalCdf(Math.abs(z)));
    p = Math.min(1, Math.max(0, p));
    return { uStatistic: u, zScore: z, pValue: p, cliffsDelta: cliffsDelta };
  }

  // ---------------------------------------------------------------- judge --

  const Classification = {
    PASS: 'Pass',
    HIGH: 'High',
    LOW: 'Low',
    NODATA: 'Nodata',
  };

  function isFailure(classification) {
    return classification === Classification.HIGH || classification === Classification.LOW;
  }

  function defaultSpec(spec) {
    return Object.assign(
      {
        direction: 'increase',
        weight: 1,
        critical: false,
        tolerance: 0.2,
        significance: 0.05,
        minSamples: 20,
      },
      spec
    );
  }

  function classifyMetric(rawSpec, canary, baseline) {
    const spec = defaultSpec(rawSpec);
    const result = {
      name: spec.name,
      weight: spec.weight,
      critical: spec.critical,
      canaryCount: canary.length,
      baselineCount: baseline.length,
      classification: Classification.NODATA,
      reason: '',
      canaryMedian: null,
      baselineMedian: null,
      canaryP95: null,
      baselineP95: null,
      pValue: null,
      cliffsDelta: null,
    };

    if (canary.length < spec.minSamples || baseline.length < spec.minSamples) {
      result.reason =
        'insufficient samples (canary=' +
        canary.length +
        ', baseline=' +
        baseline.length +
        ', required=' +
        spec.minSamples +
        ')';
      return result;
    }

    result.canaryMedian = percentile(canary, 50);
    result.baselineMedian = percentile(baseline, 50);
    result.canaryP95 = percentile(canary, 95);
    result.baselineP95 = percentile(baseline, 95);

    const test = mannWhitneyU(canary, baseline);
    result.pValue = test.pValue;
    result.cliffsDelta = test.cliffsDelta;

    const significant = test.pValue < spec.significance;
    const material = Math.abs(test.cliffsDelta) >= spec.tolerance;

    if (!significant) {
      result.classification = Classification.PASS;
      result.reason = 'no significant difference (p=' + test.pValue.toFixed(4) + ')';
      return result;
    }
    if (!material) {
      result.classification = Classification.PASS;
      result.reason =
        'significant but within tolerance (delta=' +
        test.cliffsDelta.toFixed(3) +
        ', tolerance=' +
        spec.tolerance +
        ')';
      return result;
    }

    const canaryHigher = test.cliffsDelta > 0;
    const regression =
      (spec.direction === 'increase' && canaryHigher) ||
      (spec.direction === 'decrease' && !canaryHigher) ||
      spec.direction === 'either';

    if (!regression) {
      result.classification = Classification.PASS;
      result.reason =
        'moved in the improving direction (delta=' + test.cliffsDelta.toFixed(3) + ')';
      return result;
    }

    result.classification = canaryHigher ? Classification.HIGH : Classification.LOW;
    result.reason =
      'canary ' +
      (canaryHigher ? 'higher' : 'lower') +
      ' than baseline: median ' +
      result.canaryMedian.toFixed(2) +
      ' vs ' +
      result.baselineMedian.toFixed(2) +
      ' (p=' +
      test.pValue.toFixed(4) +
      ', delta=' +
      (test.cliffsDelta >= 0 ? '+' : '') +
      test.cliffsDelta.toFixed(3) +
      ')';
    return result;
  }

  function judge(specs, canarySamples, baselineSamples, passThreshold, marginalThreshold) {
    if (!specs || !specs.length) throw new Error('canary judgment requires at least one metric');
    const pass = passThreshold === undefined ? 95 : passThreshold;
    const marginal = marginalThreshold === undefined ? 75 : marginalThreshold;

    const results = specs.map((spec) =>
      classifyMetric(spec, canarySamples[spec.name] || [], baselineSamples[spec.name] || [])
    );

    const scored = results.filter((r) => r.classification !== Classification.NODATA);
    if (!scored.length) {
      return {
        score: 0,
        verdict: 'FAIL',
        results: results,
        passThreshold: pass,
        marginalThreshold: marginal,
        summary: 'No metric produced enough data to judge; refusing to promote.',
      };
    }

    const totalWeight = scored.reduce((sum, r) => sum + r.weight, 0);
    const passingWeight = scored
      .filter((r) => !isFailure(r.classification))
      .reduce((sum, r) => sum + r.weight, 0);
    const score = totalWeight ? (100 * passingWeight) / totalWeight : 0;

    const criticalFailures = results.filter((r) => isFailure(r.classification) && r.critical);

    let verdict;
    let summary;
    if (criticalFailures.length) {
      verdict = 'FAIL';
      summary =
        'Critical metric regression: ' +
        criticalFailures.map((r) => r.name).join(', ') +
        '. Score ' +
        score.toFixed(1) +
        ' is not consulted.';
    } else if (score >= pass) {
      verdict = 'PASS';
      summary = 'Score ' + score.toFixed(1) + ' >= pass threshold ' + pass + '.';
    } else if (score >= marginal) {
      verdict = 'MARGINAL';
      summary =
        'Score ' +
        score.toFixed(1) +
        ' sits between marginal (' +
        marginal +
        ') and pass (' +
        pass +
        ') thresholds; human judgment required.';
    } else {
      verdict = 'FAIL';
      summary = 'Score ' + score.toFixed(1) + ' < marginal threshold ' + marginal + '.';
    }

    const nodata = results.filter((r) => r.classification === Classification.NODATA);
    if (nodata.length) {
      summary += ' Excluded ' + nodata.length + ' metric(s) with insufficient data.';
    }

    return {
      score: score,
      verdict: verdict,
      results: results,
      passThreshold: pass,
      marginalThreshold: marginal,
      summary: summary,
    };
  }

  // ------------------------------------------------------------ simulation --

  /* Deterministic RNG so a given set of slider positions always produces the
     same verdict. A simulator that gives different answers on every click
     teaches nothing. */
  function mulberry32(seed) {
    let a = seed >>> 0;
    return function () {
      a |= 0;
      a = (a + 0x6d2b79f5) | 0;
      let t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function gaussPair(rng) {
    let u = 0;
    let v = 0;
    while (u === 0) u = rng();
    while (v === 0) v = rng();
    return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
  }

  /* Mirrors services/paved-road-demo/app.py: lognormal latency, Bernoulli
     errors, saturation tracking configured latency. */
  function simulate(options) {
    const rng = mulberry32(options.seed);
    const windows = options.windows;
    const perWindow = options.requestsPerWindow;

    const latency = [];
    const errorRate = [];
    const throughput = [];
    const saturation = [];

    for (let w = 0; w < windows; w++) {
      let errors = 0;
      let windowLatency = 0;
      for (let r = 0; r < perWindow; r++) {
        const tail = Math.exp(gaussPair(rng) * 0.45);
        const ms = Math.max(0.5, options.latencyMs * tail + rng() * 6);
        latency.push(ms);
        windowLatency += ms;
        if (rng() < options.errorRate) errors++;
      }
      errorRate.push((100 * errors) / perWindow);
      const meanMs = windowLatency / perWindow;
      throughput.push((perWindow * 1000) / (meanMs * perWindow / options.concurrency));
      saturation.push(Math.min(99, Math.max(1, options.latencyMs * 1.4 + (rng() * 6 - 3))));
    }

    const series = {};
    series.request_latency_ms = latency;
    series.error_rate_pct = errorRate;
    series.throughput_rps = throughput;
    series.cpu_saturation_pct = saturation;
    return series;
  }

  const api = {
    fractionalRanks: fractionalRanks,
    normalCdf: normalCdf,
    percentile: percentile,
    mannWhitneyU: mannWhitneyU,
    classifyMetric: classifyMetric,
    judge: judge,
    simulate: simulate,
    Classification: Classification,
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = api;
  } else {
    root.GoldenPath = api;
  }
})(typeof globalThis !== 'undefined' ? globalThis : this);
